hathor/utils/zope.py

```diff
@@ -0,0 +1,39 @@
+#  Copyright 2023 Hathor Labs
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from typing import Any, Literal, TypeVar, Union, cast
+
+from zope.interface import Interface
+from zope.interface.exceptions import Invalid
+from zope.interface.verify import verifyObject
+
+T = TypeVar('T', bound=Interface)
+
+
+def verified_cast(interface_class: type[T], obj: Any) -> Union[T, Literal[False]]:
+    """
+    Receive a zope interface and an object, and return a cast to this interface if the object implements it.
+    Return False otherwise.
+
+    Note: this return type loosely simulates what would be a Result type in other languages.
+    """
+    result: Union[T, Literal[False]] = False
+
+    try:
+        if verifyObject(interface_class, obj):
+            result = cast(T, obj)
+    except Invalid:
```

Comment:
Why not just raise the exception?

Reply:
Because I don't want it to fail in the logic used by the new PubSub implementation, instead I'm just checking the cast.

I have implemented a new utils function that uses this one and raises if the cast fails in https://github.com/HathorNetwork/hathor-core/pull/694/commits/ddcde1028e1c8866b376a185a5ae74b4e80348f4.

Reply:
Got it. Thanks for the clarification. I still think that returning `None` and then asserting it's not `None` has the same effect then just not catching the exception. Anyway, the implementation seems good to me now. Thanks!

Reply:
I wanted to return `None` instead of raising the exception because it simplifies this code in the PubSub:

```python
if not isInIOThread() and (threaded_reactor := verified_cast(IReactorFromThreads, self.reactor)):
    ...
```

By doing it like this, I don't need a `try-except` block in the PubSub.

Reply:
Got it. Thanks again for the explanation.